src/workerd/api/basics.c++

```diff
@@ -696,6 +817,132 @@ void AbortSignal::triggerAbort(
+}
+
+jsg::Ref<AbortSignal> AbortSignal::deserialize(
+    jsg::Lock& js, rpc::SerializationTag tag, jsg::Deserializer& deserializer) {
+  auto& handler = KJ_REQUIRE_NONNULL(
+      deserializer.getExternalHandler(), "got AbortSignal on non-RPC serialized object?");
+  auto externalHandler = dynamic_cast<RpcDeserializerExternalHandler*>(&handler);
+  KJ_REQUIRE(externalHandler != nullptr, "got AbortSignal on non-RPC serialized object?");
+
+  auto isCanceled = static_cast<bool>(deserializer.readRawUint32());
+  auto flag = static_cast<Flag>(deserializer.readRawUint32());
+  auto reason = deserializer.readValue(js);
+
+  if (isCanceled) {
+    // The signal is already aborted and cannot be triggered again. We don't need to set up RPC.
+    return abort(js, reason);
+  }
+
+  if (flag == Flag::NEVER_ABORTS) {
+    // The signal can't be aborted. We don't need to setup RPC
+    return js.alloc<AbortSignal>(/* exception */ kj::none, /* maybeReason */ kj::none, flag);
+  }
+
+  auto reader = externalHandler->read();
+  KJ_REQUIRE(reader.isAbortTrigger(), "external table slot type does't match serialization tag");
+
+  // The AbortSignalImpl will receive any remote triggerAbort requests and fulfill the promise with the reason for abort
+
+  auto signal = js.alloc<AbortSignal>(/* exception */ kj::none, /* maybeReason */ kj::none, flag);
+
+  auto paf = kj::newPromiseAndFulfiller<void>();
+  externalHandler->setLastStream(
+      kj::heap<AbortTriggerRpcServer>(kj::mv(paf.fulfiller), kj::addRef(*signal->pendingReason)));
+  signal->maybeRpcAbortPromise = kj::mv(paf.promise);
+
+  return signal;
+}
+
+void AbortSignal::skipReleaseForTest() {
+  for (auto& cap: rpcClients) {
+    cap->skipReleaseForTest = true;
+  }
+
+  rpcClients.clear();
+}
+
+kj::Promise<void> AbortSignal::sendToRpc(kj::Array<kj::byte>&& reason) {
+  auto& ioContext = IoContext::current();
+
+  KJ_IF_SOME(outputLocks, ioContext.waitForOutputLocksIfNecessary()) {
+    co_await outputLocks;
+  }
+
+  kj::Vector<kj::Promise<void>> promises;
+  for (auto& cap: rpcClients) {
+    promises.add(cap->abort(reason));
+  }
+
+  co_await kj::joinPromises(promises.releaseAsArray());
+}
+
+bool AbortSignal::hasPendingReason() {
+  return pendingReason->getWrapped() != nullptr;
+}
+
+kj::Maybe<jsg::JsValue> AbortSignal::deserializePendingReason(jsg::Lock& js) {
+  if (!hasPendingReason()) {
+    // Pending reason not initialized; abort not yet triggered
+    return kj::none;
+  }
+
+  KJ_SWITCH_ONEOF(pendingReason->getWrapped()) {
+    KJ_CASE_ONEOF(v8Serialized, kj::Array<kj::byte>) {
+      jsg::Deserializer des(js, v8Serialized);
+      return kj::some(des.readValue(js));
+    }
+
+    KJ_CASE_ONEOF(exception, kj::Exception) {
+      return kj::some(js.exceptionToJsValue(kj::cp(exception)).getHandle(js));
+    }
+  }
+
+  KJ_UNREACHABLE;
+}
+
+void AbortSignal::subscribeToRpcAbort(jsg::Lock& js) {
+  KJ_IF_SOME(abortPromise, maybeRpcAbortPromise) {
```

Comment:
Maybe worth a comment like:

> For an AbortSignal received over RPC, the first time someone registers an event on the signal, we want to arrange to awaitIo() for the underlying RPC signal. If no one is actually listening, though, we don't want to awaitIo() since it blocks hibernation in actors.